proposals/003-metric-improvements.md

```diff
@@ -0,0 +1,269 @@
+| kroxylicious_inbound_downstream_decoded_messages | Counter      | flowing, virtualCluster                     | Incremented by one every time every inbound RPC arriving _from the downstream that the proxy needs to decode_. (*1) |
+| kroxylicious_payload_size_bytes                  | Distribution | flowing, virtualCluster, ApiKey, ApiVersion | Incremented with RPC's frame size for every _decoded_ RPC from either the _downstream or upstream_. (*2)            |
+| kroxylicious_downstream_connections              | Counter      | flowing, virtualCluster                     | Incremented by one each time a connection arrives from the downstream                                               |
+| kroxylicious_downstream_errors                   | Counter      | flowing, virtualCluster                     | Incremented by one each time a connection fails owning to a downstream error (dropped connection) (*3)              |
+| kroxylicious_upstream_connections_attempts       | Counter      | flowing, virtualCluster                     | Incremented by one each time a connection attempt to made to the upstream.                                          |
+| kroxylicious_upstream_errors                     | Counter      | flowing, virtualCluster                     | Incremented by one each time a connection attempt fails owning to a upstream error (tls negotiation etc)            |
+
+### Specific short-comings/weaknesses
+
+* *1 - these metrics record only the **requests**, there's nothing analogous that records the responses.
+* *1 - these metrics aren't discriminated by ApiKey (etc.). This limits their usefulness for problem-solving.
+* *2 - the metric name is quite misleading - nothing tells the user that it is only recording the sizes of RPCs that are actually **decoded**.
+* *2 - there are no quantised buckets configured for this metric. it gives the user count/sum/max values only. it would be
+  hard to use this metric to understand something about produce requests or fetch responses.
+* *1 / *2 - we have contradictory meanings for the `flowing` label.  in *2, requests flow `UPSTREAM`, responses flow `DOWNSTREAM`.
+  In *1 requests flow `DOWNSTREAM` which makes no sense.
+* *3 - downstream errors does not record account for errors such as TLS negotiation errors or failures to determine virtual cluster or broker.  
+
+### General short-comings/weaknesses
+
+* The metrics endpoint populates `HELP` for each metric, but this is populated with nothing more than the metric's name.
+  We ought to be giving a clear description of the metric.
+* Some of the label names don't follow Prometheus naming conventions.  Prometheus recommends snake case whereas we have used camel case in some cases.
+* Some of the metrics names aren't too clear.  For instance, the word payload has defined meaning in the [Kafka Protocol Guide](https://kafka.apache.org/protocol).
```

Comment:
I can't spot `payload` mentioned in the protocol doc linked, there are a couple of mentions in the kafka doc though:

```
design.html
565:    Compaction also allows for deletes. A message with a key and a null payload will be treated as a delete from the log. Such a record is sometimes referred to as a <i>tombstone</i>. This delete marker will cause any prior message with that key to be removed (as would any new
```

and a few usages of payload in the kafka clients code, so yeah it's overloaded

Reply:
> 565:    Compaction also allows for deletes. A message with a key and a null payload will be treated...

ha - that's wrong.   They actually mean `value`.